Optimized EMA implementation (#94820)
pytorch/pytorch: docs/source/optim.rst

    lr_scheduler.OneCycleLR
    lr_scheduler.CosineAnnealingWarmRestarts

Weight Averaging (SWA and EMA)
------------------------------

:mod:`torch.optim.swa_utils` implements Stochastic Weight Averaging (SWA) and Exponential Moving Average (EMA). In particular,

> :class:`torch.optim.swa_utils.AveragedModel` class implements SWA and EMA models,

Reviewer: nit: either 
```suggestion
the :class:`torch.optim.swa_utils.AveragedModel` class implements SWA and EMA models,
```
or
```suggestion
:class:`torch.optim.swa_utils.AveragedModel` implements SWA and EMA models,
```

Author: Fixed

:class:`torch.optim.swa_utils.SWALR` implements the SWA learning rate scheduler and
:func:`torch.optim.swa_utils.update_bn` is a utility function used to update SWA/EMA batch
normalization statistics at the end of training.

> SWA has been proposed in `Averaging Weights Leads to Wider Optima and Better Generalization`_.

Reviewer: Somewhere in the docs should motivate EMA as well.

    lr_scheduler.OneCycleLR
    lr_scheduler.CosineAnnealingWarmRestarts

Weight Averaging (SWA and EMA)
------------------------------

:mod:`torch.optim.swa_utils` implements Stochastic Weight Averaging (SWA) and Exponential Moving Average (EMA). In particular,
the :class:`torch.optim.swa_utils.AveragedModel` class implements SWA and EMA models,
:class:`torch.optim.swa_utils.SWALR` implements the SWA learning rate scheduler and
:func:`torch.optim.swa_utils.update_bn` is a utility function used to update SWA/EMA batch
normalization statistics at the end of training.

SWA has been proposed in `Averaging Weights Leads to Wider Optima and Better Generalization`_.

EMA is a widely known technique to reduce the training time by reducing the number of weight updates needed. It is a variation of `Polyak averaging`_.

.. _`Averaging Weights Leads to Wider Optima and Better Generalization`: https://arxiv.org/abs/1803.05407

> .. _`Polyak averaging`: https://epubs.siam.org/doi/abs/10.1137/0330046

Author: Yes polyak uses equal weights. Added some clarification and fixed the reference.

Constructing averaged models
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The `AveragedModel` class serves to compute the weights of the SWA or EMA model.

You can create an SWA averaged model by running:

>>> averaged_model = AveragedModel(model)

EMA models are constructed by specifying the ``multi_avg_fn`` argument as follows:

>>> decay = 0.999
>>> averaged_model = AveragedModel(model, multi_avg_fn=get_ema_multi_avg_fn(decay))

Decay is a parameter between 0 and 1 that controls how fast the averaged parameters are decayed. If not provided to ``get_ema_multi_avg_fn``, the default is 0.999.

``get_ema_multi_avg_fn`` returns a function that applies the following EMA equation to the weights:

.. math:: W^\textrm{EMA}_{t+1} = \alpha W^\textrm{EMA}_{t} + (1 - \alpha) W^\textrm{model}_t

where alpha is the EMA decay.

Here the model ``model`` can be an arbitrary :class:`torch.nn.Module` object. ``averaged_model``
will keep track of the running averages of the parameters of the ``model``. To update these

> averages, you should use the :func:`update_parameters` function:

Author: Fixed

Constructing averaged models
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The `AveragedModel` class serves to compute the weights of the SWA or EMA model.

You can create an SWA averaged model by running:

>>> averaged_model = AveragedModel(model)

EMA models are constructed by specifying the ``multi_avg_fn`` argument as follows:

> >>> averaged_model = AveragedModel(model, multi_avg_fn=get_ema_multi_avg_fn(decay))

Reviewer: Should we also note that decay is something the user has to provide? Better yet, could it be made optional where if it's not provided, decay will default to a certain value?

Author: Added default value and formula

Here the model ``model`` can be an arbitrary :class:`torch.nn.Module` object. ``averaged_model``
will keep track of the running averages of the parameters of the ``model``. To update these

> averages, you can use the :func:`update_parameters` function:

Reviewer: This is not the fault of this PR, but since we are here, I might as well point it out: it is unclear whether it is intended for the user to call update_parameters and when. From the usage below, it looks like it should be called after the optimizer.step() after a specified step number (swa_start) and it would be good to clarify this usage.

Author: Added a sentence to clarify

>>> averaged_model.update_parameters(model)

Custom averaging strategies
^^^^^^^^^^^^^^^^^^^^^^^^^^^

By default, :class:`torch.optim.swa_utils.AveragedModel` computes a running equal average of
the parameters that you provide, but you can also use custom averaging functions with the
``avg_fn`` or ``multi_avg_fn`` parameters:

- ``avg_fn`` allows defining a function operating on each parameter tuple (averaged parameter, model parameter) and should return the new averaged parameter.

> - ``multi_avg_fn`` allows defining more efficient operations acting on multiple (averaged parameter, model parameter) tuples at the same time, for example using the ``torch._foreach*`` functions. This function should update the averaged parameters in-place.

Reviewer: ```suggestion
- ``multi_avg_fn`` allows defining more efficient operations acting on a tuple of parameter lists, (averaged parameter list, model parameter list), at the same time, for example using the ``torch._foreach*`` functions. This function must update the averaged parameters in-place.
```
It is more accurate to say the function operates on a tuple of lists instead of lists of tuples, and we should be clear that we expect inplace updates.

Author: Fixed
